docs/proposals/20210222-azure-machinepool-machine.md

```diff
@@ -0,0 +1,332 @@
+    // for logging and human consumption.
+    //
+    // Any transient errors that occur during the reconciliation of MachinePools
+    // can be added as events to the MachinePool object and/or logged in the
+    // controller's output.
+    // +optional
+    FailureMessage *string `json:"failureMessage,omitempty"`
+    
+    // Conditions defines current service state of the AzureMachinePool.
+    // +optional
+    Conditions clusterv1.Conditions `json:"conditions,omitempty"`
+    
+    // LongRunningOperationState saves the state for an Azure long running operations so it can be continued on the
+    // next reconciliation loop.
+    // +optional
+    LongRunningOperationState *infrav1.Future `json:"longRunningOperationState,omitempty"`
+    
+    // LatestModelApplied indicates the instance is running the most up-to-date VMSS model. A VMSS model describes
+    // the image version the VM is running. If the instance is not running the latest model, it means the instance
+    // may not be running the version of Kubernetes the Machine Pool has specified and needs to be updated.
+    LatestModelApplied bool `json:"latestModelApplied"`
+    
+    // Ready is true when the provider resource is ready.
+    // +optional
+    Ready bool `json:"ready"`
+}
+```
+
+#### Proposed Controller Changes
+
+* Create a new AzureMachinePoolMachine controller.
+* Remove VMSS instance status tracking logic from AzureMachinePool controller and moving it to AzureMachinePoolMachine
+  controller.
+* (optional) Introduce rate limiting behavior to AzureMachinePool* controllers to ensure Azure API limits are not 
+  exceeded.  
+
+#### Proposed Changes of Responsibility
+Currently in CAPZ, the AzureMachinePool controller is responsible for both the Virtual Machine Scale Set (VMSS) and the
+instances created by the VMSS. The proposed change would separate the responsibility of managing the state of the VMSS
+and the instances created by the VMSS. This would introduce a new AzureMachinePoolMachine controller and a new
+MachinePoolMachineScope. The responsibilities would be as follows.
+
+**AzureMachinePool Responsibilities:**
+- Create AzureMachinePoolMachine instances when a new VMSS instance is observed. The AzureMachinePoolMachine spec should
+  have the `ProviderID` field set with the observed resource ID. The AzureMachinePool should also be added to the
+  AzureMachinePoolMachine's OwnerReferences. 
+- Upgrade / replace selection of AzureMachinePoolMachine instances. When a change to the AzureMachinePool model occurs,
+  the MachinePoolScope will be responsible for coordinating the rollout of the updated model by selecting
+  AzureMachinePoolMachines to delete / replace with respect to MaxUnavailable.
+- Scale up. AzureMachinePool should increase the number of VMSS replicas if the replica count increases on MachinePool
+- Scale down. AzureMachinePool should select and delete AzureMachinePoolMachines that are overprovisioned with respect
```

Comment:
> should select

how does selection happen? at random? oldest first? newest first?

Reply:
Added to take into account the DeletePolicy.